### PR TITLE
service: prepare for epoch support in GrpcServer

### PR DESCRIPTION
**Pull request by benesch**
Only the last two commits. The first two commits are #13388.

----

This commit reworks the concurrency setup in `GrpcServer`. The new
implementation directly uses the task created by the gRPC server to
forward commands to the underlying client, with a straightforward
cancellation protocol to fence out a pre-existing client.

This results in an implementation whose correctness is a bit easier (for
me) to reason about, and leaves a natural place to slot in epoch
checking once we tackle https://github.com/MaterializeInc/materialize/issues/13377.

As part of this work, remove the `--linger` and `--reconcile` options.
These are required for correct functioning in Materialize Cloud, so
better not to have it optional locally, IMO. The `--reset` option to
`bin/environmentd` is updated to find and kill any orphan `storaged` and
`computed` processes, which should help mitigate any bugs in the
reconciliation logic.

Touches https://github.com/MaterializeInc/materialize/issues/13377.

### Motivation

   * This PR refactors existing code.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -  n/a

### CODE REVIEW

**Comment by benesch** [comment]
This is ready for review. I had to fix a small bug in `ComputeCommandReconcile` in which `recv` would occasionally return `Ok(None)`, which would incorrectly indicate to higher layers of the client stack that the replica had gracefully terminated. The fix is simply to add a loop inside `ComputeCommandReconcile::recv` to avoid ever returning `None`.

**Comment by benesch** [comment]
I realize now that enabling reconciliation by default is the last step of #10549, but I'm proposing to do it here. I think that reordering is worthwhile. We're looking to spend the month of July running some permanent environments in Materialize Cloud for the DevEx team; better to smoke out the bugs early.

File: src/orchestrator-process/src/lib.rs [hunk 1]
```diff
@@ -191,6 +191,7 @@ impl NamespacedOrchestrator for NamespacedProcessOrchestrator {
                                     port_metadata_file_location,
                                 ),
                             })));
+                            processes.push(port_metadata);
```
Comment (antiguru) [inline src/orchestrator-process/src/lib.rs:194]: Oops! :)

**Review by antiguru — APPROVED** [review]
Nice

**Review by lluki — APPROVED** [review]
LGTM

**Comment by benesch** [comment]
Thanks for the reviews! I checked with @brennanvincent on Slack and he's happy to review this post merge. I also tested this manually on cloud and it worked even with large cluster sizes.

**Pull request merged**